### PR TITLE
Add support for the Nordic nRF9160

### PR DESCRIPTION
**Pull request by jonathanpallant**
This PR adds support for the Nordic nRF9160. As the chip is so similar to the nRF52, I have added support there, but we may want to rename the shared crate to `nrf-hal-common`.

The changes were largely:

* Adding a feature to enable 9160 support to nrf52-hal-common
* Adding nrf9160-hal which uses nrf52-hal-common
* Adding nRF9160-DK, which is a board support package for the PCA10090 nRF9160 developer kit
* A fix to the UARTE0 code, without which my buffers did not complete transmitting, giving partial output.
* An extra Timer API I found useful for my demo

The HAL only supports peripherals in Non-Secure mode, as that's the most common usecase. Support for secure mode could be added later. I elected to import the `xxx_NS` peripherals as `xxx` to reduce the changes required to the rest of the driver (which otherwise seem to work unchanged).

I have tested the GPIO and the UART in a separate application which is not yet public.

These changes are Copyright (c) 42 Technology Ltd, as noted in the crate metadata. I hereby confirm I am authorised by 42 Technology Ltd to release these changes under both the Apache 2.0 and MIT licences, in-line with the rest of the software in this repository.

### CODE REVIEW

**Comment by Yatekii** [comment]
Wow, this is great, thanks a lot!
I'll have a closer look later.
Your suggestion to maybe branch stuff out into a common crate repo makes sense to me. We already discussed this some times and I think now it might be time to do this.

**Comment by jonathanpallant** [comment]
I think there is merit in keeping the `nrfXXX-hal` HAL crates and `nrf52-common-hal` in the same repo, as it makes it easier to make atomic changes to both at once. I just think that `nrf52-common-hal`, and the repo itself, now have the wrong name :)

**Comment by hannobraun** [comment]
Wow, this is awesome!

I don't have time to review now, but feel free to ping me, in case @Yatekii doesn't find the time soon either :-)

**Comment by thejpster** [comment]
Tagging this on my personal account, and also giving @Yatekii and @hannobraun a gentle prod ;)

File: boards/nRF9160-DK/Cargo.toml [hunk 1]
```diff
@@ -1,9 +1,23 @@
 [package]
-name = "nRF9160-DK"
+authors = [
+   "Ruben Paz <me@ruben.io>",
+   "Jonathan Pallant <jonathan.pallant@42technology.com>",
+]
```
Comment (hannobraun) [inline boards/nRF9160-DK/Cargo.toml]: List of authors seems copy-pasted. Should probably be @jonathanpallant.

File: nrf52-hal-common/src/uarte.rs [hunk 2]
```diff
@@ -116,11 +124,20 @@ impl<T> Uarte<T> where T: Instance {
             return Err(Error::TxBufferTooLong);
         }
 
+        // We can only DMA out of RAM
+        if !crate::slice_in_ram(tx_buffer) {
+            return Err(Error::BufferNotInRAM);
+        }
```
Comment (hannobraun) [inline nrf52-hal-common/src/uarte.rs]: @simonsso You spoke out against this approach in #37. I still think returning the error is clearly the best approach and would like to merge this as-is, but didn't want to sneak this by you.

File: examples/twi-ssd1306/Cargo.toml [hunk 3]
```diff
@@ -1,7 +1,11 @@
 [package]
 name = "twi-ssd1306"
 version = "0.1.0"
-authors = ["James Waples <james@wapl.es>", "Ruben Paz <me@ruben.io>"]
+authors = [
+   "James Waples <james@wapl.es>",
+   "Ruben Paz <me@ruben.io>",
+   "Jonathan Pallant <jonathan.pallant@42technology.com>"
```
[the remark this answers is not in the document]
Comment (jonathanpallant) [reply examples/twi-ssd1306/Cargo.toml]: vvv I added nrf9160 support to the features list below

File: nrf9160-hal/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+version = "0.1.0"
+description = "HAL for nRF9160 system-in-package"
+
+repository = "https://github.com/nrf-rs/nrf52-hal"
+authors = [
+    "Jonathan Pallant (42 Technology) <jonathan.pallant@42technology.com>",
+    "James Munns <james@onevariable.com>",
+    "Hanno Braun <hanno@braun-robotics.com>",
+    "John Scarrott <johnps@outlook.com>",
+    "Wez Furlong <wez@wezfurlong.org>",
```
[the remark this answers is not in the document]
Comment (jonathanpallant) [reply nrf9160-hal/Cargo.toml]: This whole project was a copy-paste of the nRF52-DK BSP so I took the authors list and added to it. If my changes have entirely removed someone else's work then we could clean up the copyright list, but is it worth the effort of checking?

[the remark this answers is not in the document]
Comment (jonathanpallant) [reply nrf9160-hal/Cargo.toml]: Again, this was a copy of another -hal so I copied over the authors list.

**Comment by hannobraun** [comment]
@jonathanpallant Thanks for the replies. The authors list is usually not used as a list of copyright holders. The current lists certainly don't reflect the full list of copyright holders, you'd have to figure those out from the Git history.

The authors are directly listed on crates.io, and if you look at the most popular crates, most have only 1 or 2 authors listed, despite surely having many more contributors. My expectation would be that only significant contributors would be added to this list. I'd certainly be surprised to receive email about a crate that I've never touched (but that copies a few lines of my code), because someone found my email on crates.io.

All that said, we've never decided on a policy, and I don't care much one way or the other. I figured I'd state my view on things, but I'll not hold it against you if you ignore me completely :-)

File: nrf9160-hal/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Yatekii) [reply nrf9160-hal/Cargo.toml]: I think it's ok. It's not up2date anyways I think. I didn't add myself and I did most of stuff for the nRF52840. I just don't care enough :)
So not a deal breaker for me.

**Comment by jonathanpallant** [comment]
Good to know! Happy to clean these up, as you are right, I don't expect the others to want to handle complaints about these crates :)

**Comment by Yatekii** [comment]
LGTM! Thanks a lot for your contribution :)

**Pull request merged**